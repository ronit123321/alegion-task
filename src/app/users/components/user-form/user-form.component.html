<form [formGroup]="form">
  <input type="text" placeholder="First Name" formControlName="first_name">
  <input type="text" placeholder="Last Name" formControlName="last_name">
  <button (click)="onSave()">Save</button>
</form>