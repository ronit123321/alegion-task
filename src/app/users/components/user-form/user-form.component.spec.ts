import { SimpleChange } from '@angular/core';
import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { User } from '@state/user/user.model';
import { UserFormComponent } from './user-form.component';

function newEvent(eventName: string, bubbles = false, cancelable = false) {
  const evt = document.createEvent('CustomEvent');
  evt.initCustomEvent(eventName, bubbles, cancelable, null);
  return evt;
}

describe('UserFormComponent', () => {
  let component: UserFormComponent;
  let fixture: ComponentFixture<UserFormComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [UserFormComponent],
      imports: [FormsModule, ReactiveFormsModule]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(UserFormComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should patch values into the form', () => {
    const user = {
      id: 8,
      email: "lindsay.ferguson@reqres.in",
      first_name: "Lindsay",
      last_name: "Ferguson",
      avatar: "https://s3.amazonaws.com/uifaces/faces/twitter/araa3185/128.jpg"
    };

    component.ngOnChanges({
      user: new SimpleChange(null, user, true)
    });

    expect(component.form.value).toEqual({
      first_name: user.first_name,
      last_name: user.last_name
    });
  });

  it('should emit the userChange event when submitted', () => {
    const user = {
      id: 8,
      email: "lindsay.ferguson@reqres.in",
      first_name: "Lindsay",
      last_name: "Ferguson",
      avatar: "https://s3.amazonaws.com/uifaces/faces/twitter/araa3185/128.jpg"
    };
    const first_name = 'Brian';
    const first_nameDebugEl = fixture.debugElement.query(
      By.css('input[formControlName="first_name"]')
    );
    const first_nameEl = first_nameDebugEl.nativeElement as HTMLInputElement;
    const buttonDebugEl = fixture.debugElement.query(By.css('button'));

    fixture.detectChanges();

    let updatedUser: User;
    component.userChange.subscribe(u => (updatedUser = u));

    component.user = user;
    component.ngOnChanges({
      user: new SimpleChange(null, user, true)
    });

    first_nameEl.value = first_name;
    first_nameEl.dispatchEvent(newEvent('input'));

    buttonDebugEl.triggerEventHandler('click', null);

    expect(updatedUser).toEqual({
      ...user,
      first_name
    });
  });
});
