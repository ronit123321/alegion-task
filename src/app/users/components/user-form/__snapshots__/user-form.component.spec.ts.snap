// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserFormComponent should match snapshot 1`] = `
<app-user-form
  form={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  userChange={[Function EventEmitter]}
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <input
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="first_name"
      ng-reflect-name="first_name"
      placeholder="First Name"
      type="text"
    />
    <input
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="last_name"
      ng-reflect-name="last_name"
      placeholder="Last Name"
      type="text"
    />
    <button>
      Save
    </button>
  </form>
</app-user-form>
`;
